subsys/bluetooth/host/gatt.c

```diff
@@ -3523,16 +3524,25 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
 		return false;
 	}
 
-	ccc = attr->user_data;
+	if (!attr->read) {
+		LOG_ERR("Read method not set");
+		return false;
+	}
+
+	len = attr->read(conn, attr, ccc_bits_encoded, sizeof(ccc_bits_encoded), 0);
+	if (len < 0) {
+		LOG_ERR("Failed to read attribute %p (err %zd)", attr, len);
+		return false;
+	} else if (len != sizeof(ccc_bits_encoded)) {
+		LOG_ERR("Invalid read length: %zd", len);
+		return false;
+	}
 
-	/* Check if the connection is subscribed */
-	for (size_t i = 0; i < BT_GATT_CCC_MAX; i++) {
-		const struct bt_gatt_ccc_cfg *cfg = &ccc->cfg[i];
+	ccc_bits = sys_get_le16(ccc_bits_encoded);
 
-		if (bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer) &&
-		    (ccc_type & ccc->cfg[i].value)) {
-			return true;
-		}
+	/* Check if the CCC bits match the subscription type */
+	if (ccc_bits & ccc_type) {
```

Comment:
The read function for CCCs (included in the ccc declaration here: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/gatt.h#L1045), leads to this check: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L2159-L2160, so the check is still included!

(Called here in read: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L2179)

Reply:
This is assuming that the `attr->read` is `bt_gatt_attr_read_ccc`, isn't it? In your previous PR you removed the assumption that the GATT macros were used, but here you seem to rely on `BT_GATT_CCC_MANAGED` being used. If an application implement its own version of `bt_gatt_attr_read_ccc` that does not have the `bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer)` check, then this is changing behavior. 

Based on the reasoning from https://github.com/zephyrproject-rtos/zephyr/pull/80570 I think it is OK, but we are expecting that users either use the GATT macros (which contradicts the need for https://github.com/zephyrproject-rtos/zephyr/pull/80570) or that the users uses `bt_gatt_attr_read_ccc` or implements their own function that ensure this check. None of that is documented, and if anyone relied on the check here instead of having it in their own read function, then their software may start to silently break. 

TL;DR: It is OK to not have the check here, but we should probably mention it in the release notes or migration guides, as this is changing the behavior of this function based on undocumented expectations.

Reply:
> This is assuming that the `attr->read` is `bt_gatt_attr_read_ccc`, isn't it?

No, it does not care whether `attr->read` is `bt_gatt_attr_read_ccc`.

All we care about is that `attr->uuid` is `BT_UUID_GATT_CCC`. Then `attr->read` must act according to spec and give the value of the CCC.

In GATT, a remote client can see if it's subscribed by reading the CCC value with a normal ATT read. So to see whether a remote client is subscribed, we just emulate a remote read using `attr->read(conn)`, and we immediately get the answer we want.

Reply:
Fair point!
I added a section about this to the migration guide, though I am a bit uncertain on how much we should recommend adding this check. I went for "this check should be included", but if we want a weaker or stronger recommendation I can change it :)

Reply:
There is no fair point. Please think carefully about what the "`bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer)` check" actually is.

Reply:
Answer: It part of the the lookup of `cfg`. It's meaningless by itself. It's not a "check" at all.

```c
	for (size_t i = 0; i < BT_GATT_CCC_MAX; i++) {
		const struct bt_gatt_ccc_cfg *cfg = &ccc->cfg[i];

		if (bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer)) {
			/* Found the `cfg` for `conn`. */
		}
	}
```

Reply:
Fair point @alwa-nordic 

GATT is basically underdocumented in how it works and how applications using their own functions should work, so when changing anything here it can break stuff that may have worked previously due to assumptions from applications. 

This PR is fine with or without the migration guide note, and the real issue is the lack of documentation IMO :)